p4-16/psa/PSA.mdk

```diff
@@ -1288,45 +1288,199 @@ to generate values.
-profiles, with additional support to define groups of
-entries. Action selectors require a hash algorithm to select
-members in a group. The only data plane operation required is to
-instantiate this extern. When the control plane adds entries
-(members) into the extern, they are essentially populating the
-corresponding table entries.
+Action selectors implement yet another mechanism to populate table
+entries with action specifications that have been defined outside the
+table entry. They are more powerful than action profiles because they also
+provide the ability to dynamically select the action specification to apply
+upon maching a table entry. An action selector extern can be instantiated as
+a resource in the P4 program, similar to action profiles. Furthermore,
+a table that uses this action selector must specify its implementation attribute
+as the action selector instance.
+
+~ Figure { #fig-as; caption: "Action selectors in PSA"; page-align: here; }
+![as]
+~
+[as]: action_selector.png { width: 90%; }
+
+Figure [#fig-as] illustrates a table that has an action selector implementation.
+In this example, the table has a match key consisting of an LPM on header field
+`h.f`. A second match type `selector` is used to define the fields that are
+used to look up the action specification from the selector at runtime.
+
+A table with an action action selector implementation consists of entries that
+point to either an action profile member reference or an action profile group
+reference. An action selector instance can be logically visualized as two tables
+as shown in Figure [#fig-as]. The first table contains a mapping from group
+references to a set of member references. The second table contains a mapping
+from member references to action specifications.
+
+When a packet matches a table entry at runtime, the controller-assigned
+reference of the action profile member or group is read. If the entry points to
+a member then the corresponding action specification is applied to the packet.
+However, if the entry points to a group, a dynamic selection algorithm is used
+to select a member from the group, and the action specification
+corresponding to that member is applied. The dynamic selection algorithm is
+specified as a parameter when instantiating the action selector.
+
+Action selector members may only specify action types defined
+in the `actions` attribute of the implemented table. An action selector instance
+may be shared across multiple tables only if all such tables define the same
+set of actions in their `actions` attribute. Tables with an action selector
+implementation cannot define a default action. The default action for such
+tables is implicitly set to `NoAction`.
+
+The dynamic selection algorithm requires a field list as an input for generating
+the index to a member entry in a group. This field list is created by using the
+match type `selector` when defining the table match key. The match fields of
```

Comment:
is that the long term plan for PSA? I though the consensus was to use P4 "virtual" functions. Or do we think this won't get added to the language spec in time?